### PR TITLE
xdsl: introduce {IRNode, BinaryOperation} baseclass

### PR DESCRIPTION
**Pull request by georgebisbas**
Goals:
- Reduce code duplication
- Towards avoiding cyclic imports
- Rethinking hierarchy Region->Block->Operation (is not cyclic but some properties point to sth like that)

### CODE REVIEW

File: xdsl/ir.py [hunk 1]
```diff
@@ -421,8 +446,21 @@ def verify(self, verify_nested_ops: bool = True) -> None:
             for region in self.regions:
                 region.verify()
 
+    # TODO replace with trait
     def verify_(self) -> None:
-        pass
+        try:
```
Comment (georgebisbas) [inline xdsl/ir.py:451]: Need ideas to avoid this awful hack

File: xdsl/ir.py [hunk 2]
```diff
@@ -421,8 +446,21 @@ def verify(self, verify_nested_ops: bool = True) -> None:
...
+            assert self.lhs
+        except AttributeError:
+            return
+
+        if self.lhs.typ != self.rhs.typ or self.rhs.typ != self.result.typ:
```
Comment (webmiche) [inline xdsl/ir.py:456]: I don't agree with this verifier. AFAIU, we can have operations with different input and result types, like for example a `convert` operation that converts i32 to i64 or something.

I might be missing something here feel free to point out!

Comment (webmiche) [reply xdsl/ir.py:456]: Actually, the fields `lhs` and `rhs` are not even guaranteed to be there, right? And if there is a `lhs`, this is not guaranteed to be a binary operation with the same input and result types.

**Review by webmiche — CHANGES_REQUESTED** [review]
I agree with the move to a Node class. I am not quite sure about the name though. Maybe `IRConstruct` might be more descriptive? `Node` hints strongly at a tree-based data structure, which we have in some sense (everything has exactly one parent), but I am not sure we want to sell it like this.

I also do not agree with the way you are handling verifiers though. IMO, if you want to remove them, there should be a subclass of `Operation` called `BinaryOperation` that has a `lhs` and `rhs`, which have to have the same type. This subclass can then override the `verify_` function with the invariant here. Wether this subclass should be delivered with xDSL or not needs to be discusses separately IMO.

File: xdsl/ir.py (hunk 2 above)
Comment (georgebisbas) [reply xdsl/ir.py:456]: > 

This happens in scf_ops.xdsl right?

Comment (georgebisbas) [reply xdsl/ir.py:456]: > Actually, the fields `lhs` and `rhs` are not even guaranteed to be there, right? And if there is a `lhs`, this is not guaranteed to be a binary operation with the same input and result types.

That is why I have the asserts which I consider awful

Comment (webmiche) [reply xdsl/ir.py:456]: > This happens in scf_ops.xdsl right?

Yes, there is an example of that in the `arith.cmpi` operation.

Comment (webmiche) [reply xdsl/ir.py:456]: > > Actually, the fields `lhs` and `rhs` are not even guaranteed to be there, right? And if there is a `lhs`, this is not guaranteed to be a binary operation with the same input and result types.
> 
> That is why I have the asserts which I consider awful

In the `BinaryOperation` subclass design I proposed, you won't need the assert as you can enforce the `BinaryOperation` to have the needed fields. That's another reason why I feel this design would be cleaner.

Comment (math-fehr) [reply xdsl/ir.py:459]: I agree, we should definitely not put this code here, and we should use `BinaryOperation` instead.

Comment (georgebisbas) [reply xdsl/ir.py:459]: I like the idea of BinaryOperation, thanks!

File: xdsl/ir.py [hunk 3]
```diff
@@ -524,8 +534,29 @@ def irdl_definition(cls) -> OpDef:
         ...
 
 
-@dataclass(eq=False)
-class Block:
+@dataclass
+class BinaryOperation(Operation):
```
Comment (webmiche) [inline xdsl/ir.py:538]: I think you might be able to give this class the fields, just like we do for other operations, just do not add the decorator. Then, you can get rid of the asserts in the verifier.

Comment (math-fehr) [reply xdsl/ir.py:538]: I'm not sure you want to do that, since then the fields will not have the proper types (since you might sometimes want to have an `IntegerType`, but sometimes else a `Vector`.

File: xdsl/ir.py [hunk 4]
```diff
@@ -544,6 +534,27 @@ def irdl_definition(cls) -> OpDef:
         ...
 
 
+@dataclass
```
Comment (math-fehr) [inline xdsl/ir.py:537]: I'm not sure if this would be the place to add this operation.
Maybe we should create a file like `ir_helper.py`? What do you think?
I would say that `ir.py` should only contains the necessary IR constructs.

File: xdsl/ir.py [hunk 5]
```diff
@@ -544,6 +534,27 @@ def irdl_definition(cls) -> OpDef:
...
+class BinaryOperation(Operation):
+    """A generic operation. Operation definitions inherit this class."""
+
+    # TODO replace with trait
+    def verify_(self) -> None:
+        try:
```
Comment (math-fehr) [inline xdsl/ir.py:543]: We can maybe add these asserts in the `__post_init__`, such that this will only be checked once.

File: xdsl/ir.py (hunk 3 above)
Comment (webmiche) [reply xdsl/ir.py:538]: Yeah, agreed :thinking: 
Not sure how to circumvent that. Maybe define them initially without a type constraint and them override them in the child with the type constraint?

File: xdsl/ir.py (hunk 5 above)
Comment (georgebisbas) [reply xdsl/ir.py:543]: These can be totally dropped, since we know that we are under a BinaryOperation

File: xdsl/ir.py (hunk 3 above)
Comment (georgebisbas) [reply xdsl/ir.py:538]: what fields do you mean?

File: xdsl/ir.py (hunk 4 above)
Comment (georgebisbas) [reply xdsl/ir.py:537]: Isn't an OPeration an IR Construct?

File: xdsl/ir.py (hunk 3 above)
Comment (webmiche) [reply xdsl/ir.py:538]: He means the fields you have to add, if you were to follow my proposal.

File: xdsl/ir.py (hunk 4 above)
Comment (georgebisbas) [reply xdsl/ir.py:534]: > Maybe we should create a file like ir_helper.py
It could also be just operation.py

**Comment by georgebisbas** [comment]
> Maybe `IRConstruct` might be more descriptive?

What about IRNode @math-fehr ?

**Comment by webmiche** [comment]
Why do you want the term `Node` in there? I personally like to think of our data structures in the pretty printed version, so:

```
%0: !i64 = arith.constant() ["value" = ... ]
```

I feel the term node makes people think of trees, which I would not want people to think about, though the data structure internally actually is a sort of tree.

File: xdsl/dialects/arith.py [hunk 6]
```diff
@@ -551,6 +455,9 @@ def get(operand1: Union[Operation, SSAValue],
             result_types=[IntegerType.from_width(1)],
             attributes={"predicate": IntegerAttr.from_int_and_width(arg, 64)})
 
+    def verify_(self) -> None:
```
Comment (webmiche) [inline xdsl/dialects/arith.py:458]: I would count comparisons as binary operations as well, considering they have two inputs. What exactly is a binary operation in this context? Aren't you describing a `BinarySameInputSameResultOperation`?

I think we might want to split this into several interfaces (or traits probably). The `Binary`, the `SameInputTypes`, and the `SameResultTypes` traits or something.

File: xdsl/ir.py [hunk 7]
```diff
@@ -274,7 +275,26 @@ def irdl_definition(cls) -> ParamAttrDef:
 
 
 @dataclass
-class Operation:
+class Node(object):
```
Comment (webmiche) [inline xdsl/ir.py:278]: I am still against the name `Node`. I prefer `IRConstruct` or something in this direction.

File: xdsl/ir.py (hunk 4 above)
Comment (webmiche) [reply xdsl/ir.py:534]: I think this conversation is not yet resolved. AFAIU, this kind of class is like providing a special kind of operation that users can use out of the box instead of having to define these kinds of parent classes themselves. Therefore, Mathieu would prefer to use it out of this file, that should mainly concern IR constructs and not such helper constructs.

File: xdsl/ir.py [hunk 8]
```diff
@@ -958,16 +965,40 @@ def move_blocks(self, region: Region) -> None:
...
-        return self.parent.get_toplevel_object()
 
-    def is_ancestor(self, op: Operation | Block | Region) -> bool:
-        "Returns true if the region is an ancestor of the operation, block, or region."
-        if op is self:
-            return True
-        if op.parent is None:
-            return False
-        return self.is_ancestor(op.parent)
+class DiagnosticException(Exception):
```
Comment (webmiche) [inline xdsl/ir.py:969]: Why did you decide to move this here? Circular imports?

**Comment by georgebisbas** [comment]
> Why do you want the term `Node` in there? I personally like to think of our data structures in the pretty printed version, so:
> 
> ```
> %0: !i64 = arith.constant() ["value" = ... ]
> ```
> 
> I feel the term node makes people think of trees, which I would not want people to think about, though the data structure internally actually is a sort of tree.

Usually this is the term used.
Even in : https://mlir.llvm.org/docs/LangRef/#high-level-structure we see:
_MLIR is fundamentally based on a graph-like data structure of nodes_

File: xdsl/dialects/arith.py (hunk 6 above)
Comment (georgebisbas) [reply xdsl/dialects/arith.py:458]: I like your proposed approach. One of the ideas I have is the following.
We should have a different dialect that can be called relational.py.
Have a look here if you have time: https://github.com/devitocodes/devito/blob/master/devito/types/relational.py

File: xdsl/ir.py (hunk 8 above)
Comment (georgebisbas) [reply xdsl/ir.py:969]: Yes, I moved this here or circular imports. Of course, it is not resolved. Any ideas to make it better?

**Comment by webmiche** [comment]
Yes, that seems to be the case. I am fine to use `IRNode` in that case, but I still feel we shouldn't name things by how they are implemented but rather by what kind of abstraction we want to present.

File: xdsl/dialects/arith.py (hunk 6 above)
Comment (webmiche) [reply xdsl/dialects/arith.py:458]: Yes I agree with this structure! Though notice that there are hard problems here due to our usage of decorators. AFAIK, the trait should not have a decorator atm.

**Comment by webmiche** [comment]
I feel like it might make sense to split this PR at this point. There are actually several things happening:
- introduce IRNode
- introduce traits
- merge diagnostics into ir

I think all of these points turned out to be controversial enough to warrant individual PRs, so I would prefer to have them as three individual PRs.

File: xdsl/dialects/arith.py (hunk 6 above)
Comment (georgebisbas) [reply xdsl/dialects/arith.py:459]: Maybe here you did not notice that` cmpi` is not inheriting from BinaryOperation!
It is as before, so we could skip for the moment.

**Comment by georgebisbas** [comment]
I isolated diagnostics to separate files, out of the ir.py.
Also added exceptions.py for further separation.
If you look at the PR in summary the main additions are the IRNode and BinaryOperation baseclasses. The rest stem from those additions. 
Also changed the title.
Let me know if you would still like to form separate PRs

File: xdsl/dialects/arith.py [hunk 9]
```diff
@@ -551,6 +456,9 @@ def get(operand1: Union[Operation, SSAValue],
...
+        pass
```
Comment (math-fehr) [inline xdsl/dialects/arith.py:460]: By default, the `verify_` function should do nothing, so we don't have to add it here

File: xdsl/ir.py [hunk 10]
```diff
@@ -211,6 +212,8 @@ class Attribute(ABC):
     Attributes are used to represent SSA variable types, and can be attached
     on operations to give extra information.
     """
+    is_Parametrized = False
```
Comment (math-fehr) [inline xdsl/ir.py:215]: Do we really need this?
Is this supposed to remove the `isinstance` checks? I would prefer keeping the `isinstance` personally.
For instance, this allows the type checker to understand what you are doing.

File: xdsl/ir.py [hunk 11]
```diff
@@ -274,7 +277,26 @@ def irdl_definition(cls) -> ParamAttrDef:
 
 
 @dataclass
-class Operation:
+class IRNode(object):
+    pass
```
Comment (math-fehr) [inline xdsl/ir.py:281]:
```suggestion
```

File: xdsl/ir.py [hunk 12]
```diff
@@ -274,7 +277,26 @@ def irdl_definition(cls) -> ParamAttrDef:
...
+        if op.parent is None:
+            return False
+        return cls.is_ancestor(op.parent)
+
+    def get_toplevel_object(cls: IRNode) -> IRNode:
+        """Get the operation, block, or region ancestor that has no parents."""
+        if cls.parent is None:
+            return cls
+        return cls.parent.get_toplevel_object()
+
```
Comment (math-fehr) [inline xdsl/ir.py:296]: Can we add `parent_op`, `parent_block`, and `parent_region` here? Essentially, this would be the nearest ancestor that has the right type, so that should work?

File: xdsl/ir.py (hunk 4 above)
Comment (math-fehr) [reply xdsl/ir.py:536]: Yes, could we move this to some other file?
Something like `ir_helper.py`, or `definition_utils.py`.
I would really like to have only the strict necessary in `ir.py`, to make it easier to understand the core structure of xdsl.

File: xdsl/parser.py [hunk 13]
```diff
@@ -727,21 +728,21 @@ def parse_optional_attribute(self,
 
         # Attribute with default format
         if parse_with_default_format:
-            if not issubclass(attr_def, ParametrizedAttribute):
+            if not attr_def.is_Parametrized:
```
Comment (math-fehr) [inline xdsl/parser.py:731]: I would really prefer using the `issubclass` here, since that helps the IDE to understand what is going on,
and also since this represent exactly what we are doing.

**Review by math-fehr — COMMENTED** [review]
Nice, thanks a lot!
In general, I would prefer though if we could the PRs relatively minimal (i.e., not changing all imports), but I believe that this will get better in time once we clean all the files.

I like the design idea, and `IRNode` makes complete sense to me.
My only worry here is that I would prefer to have `BinaryOperation` somewhere else in the code.

File: xdsl/ir.py (hunk 12 above)
Comment (webmiche) [reply xdsl/ir.py:296]: Note: This change will inhibit us from moving the IRNode to its own file, as we would have a circular dependency (IRNode needs Op/Block/Region for type hints, which themselves inherit from IRNode).

I think if we really want to design our hierarchy in a clean way, this might be a bad idea.

File: xdsl/dialects/arith.py (hunk 9 above)
Comment (georgebisbas) [reply xdsl/dialects/arith.py:460]: right

File: xdsl/ir.py (hunk 10 above)
Comment (georgebisbas) [reply xdsl/ir.py:215]: ok, I will revert these changes

File: xdsl/ir.py (hunk 12 above)
Comment (georgebisbas) [reply xdsl/ir.py:296]: I think this hierarchy refactor can be done in some next PR. Thoughts?

File: xdsl/parser.py (hunk 13 above)
Comment (georgebisbas) [reply xdsl/parser.py:731]: will revert

**Comment by georgebisbas** [comment]
I moved BinaryOperation to arith.py, this is where it should naturally belong.
It is a bit weird to have inheritance between files named:
ir.py - > utlitly/helper like name -> arith.py

File: xdsl/ir.py [hunk 14]
```diff
@@ -254,7 +253,7 @@ def print_parameter(data: DataElement, printer: Printer) -> None:
 @dataclass(frozen=True)
 class ParametrizedAttribute(Attribute):
     """An attribute parametrized by other attributes."""
-
+    is_Parametrized = True
```
Comment (math-fehr) [inline xdsl/ir.py:256]: I think this should still be removed

File: xdsl/ir.py (hunk 12 above)
Comment (math-fehr) [reply xdsl/ir.py:291]: Yes, we can always think of that in a next PR!
And if the plan is to move `IRNode` outside this file, then yes we can forget it!

File: xdsl/parser.py [hunk 15]
```diff
@@ -727,11 +730,6 @@ def parse_optional_attribute(self,
 
         # Attribute with default format
         if parse_with_default_format:
-            if not issubclass(attr_def, ParametrizedAttribute):
```
Comment (math-fehr) [inline xdsl/parser.py:730]: Why was this removed?
AFAIK, we need this, otherwise users will get an ugly error.

**Review by math-fehr — COMMENTED** [review]
I added 2/3 minor comments, then that should be good to merge!

**Comment by webmiche** [comment]
Moving it to arith.py makes sense for the moment. I think we will want to remove arith.py sometime in the future as we will/should be able to generate them at some point. Then, we could consider moving arith.py as an bigger example of a dialect than what we have in the notebooks. The ops are relatively straigthforward IMO.

File: xdsl/dialects/arith.py [hunk 16]
```diff
@@ -75,19 +82,27 @@ def from_int_constant(val: Union[int, Attribute],
...
                 "expect all input and result types to be equal")
 
+    def __hash__(self) -> int:
+        return id(self)
+
+
+@irdl_op_definition
+class Addi(BinaryOperation):
+    name: str = "arith.addi"
+    lhs = OperandDef(signlessIntegerLike)
```
Comment (webmiche) [inline xdsl/dialects/arith.py:102]: I think you could move these into the BinaryOperation.

File: xdsl/dialects/builtin.py [hunk 17]
```diff
@@ -232,27 +238,20 @@ def from_params(
...
 
-f16 = Float16Type()
-
-
 @irdl_attr_definition
 class Float32Type(ParametrizedAttribute, MLIRType):
     name = "f32"
 
 
-f32 = Float32Type()
```
Comment (webmiche) [inline xdsl/dialects/builtin.py:247]: This is removing functionality, i.e. accessing builtin.f32 to get an f32. We can consider removing this, though builtin.f32 might be more natural than builtin.Float32Type.

**Review by webmiche — APPROVED** [review]
Really like the new changes. I think we are reaching mergeability soon.

File: xdsl/dialects/arith.py (hunk 16 above)
Comment (math-fehr) [reply xdsl/dialects/arith.py:102]: We could indeed have a `IntegerBinaryOperation` or something like this.
In the future, I would like having something like `BinaryOperation[signlessIntegerLike]`, though this would require the merging of #210, and then other improvement (so we should not wait to have this).

File: xdsl/parser.py (hunk 15 above)
Comment (georgebisbas) [reply xdsl/parser.py:730]: It seems that no code goes through it.
Do you have an example?

File: xdsl/dialects/builtin.py (hunk 17 above)
Comment (georgebisbas) [reply xdsl/dialects/builtin.py:247]: Not sure I understand the functionality that is removed?
If so, we need a test/MFE to prove it?

Comment (webmiche) [reply xdsl/dialects/builtin.py:247]: You won't be able to access the Float32Type anymore by just calling f32. Where we actually want this/need this functionality is a different question. Maybe @math-fehr  can decide. Me personally, I don't think we need this functionality, so I am fine with these changes.

Comment (georgebisbas) [reply xdsl/dialects/builtin.py:247]: Lets wait for @math-fehr to see what to do

File: xdsl/dialects/arith.py (hunk 16 above)
Comment (webmiche) [reply xdsl/dialects/arith.py:102]: Ah yes, they have a type constraint that isn't the same everywhere, you are right!

File: xdsl/dialects/builtin.py (hunk 17 above)
Comment (math-fehr) [reply xdsl/dialects/builtin.py:247]: I personally kind of like having those aliases somewhere.
This is just to write `f32` rather than `Float32Type()`, which I think makes it nicer to read IMO.
Maybe we should put them next to each other at the end of the file?

File: xdsl/parser.py (hunk 15 above)
Comment (georgebisbas) [reply xdsl/parser.py:730]: I added a fail-in-purpose test and reverted this change. Overall we need to have tests for cases like that.
Otherwise one may assume that this is redundant piece of code.

File: xdsl/dialects/builtin.py (hunk 17 above)
Comment (georgebisbas) [reply xdsl/dialects/builtin.py:247]: ok, fixed. Maybe in some future PR we should also team-up somewhere stuff like:
```
AnyIntegerAttr, AnyFloat etc
```

Comment (math-fehr) [reply xdsl/dialects/builtin.py:247]: Yes, we should look into it!
Though we should be careful, since some of these constructs are also used in the file directly.

File: tests/printer_test.py [hunk 18]
```diff
@@ -647,3 +650,25 @@ def test_parse_dense_mlir():
     printer = Printer(stream=file, target=Printer.Target.MLIR)
     printer.print_op(module)
     assert file.getvalue().strip() == expected.strip()
+
+
+def test_foo_string():
+    """
+    Fail attribute in purpose. TOFIX
```
Comment (math-fehr) [inline tests/printer_test.py:657]: Looks good to me!
Why do we have a `TOFIX` here?
